src/mds/schedule/copySetScheduler.cpp

```diff
@@ -32,33 +32,98 @@ namespace curve {
 namespace mds {
 namespace schedule {
 int CopySetScheduler::Schedule() {
-    LOG(INFO) << "copysetScheduler begin";
+    LOG(INFO) << "schedule: copysetScheduler begin";
 
-    int res = 0;
+    int oneRoundGenOp = 0;
     for (auto lid : topo_->GetLogicalpools()) {
-        res = DoCopySetSchedule(lid);
+        oneRoundGenOp += DoCopySetSchedule(lid);
     }
-    return res;
+
+    LOG(INFO) << "schedule: copysetScheduler end, generate operator num "
+              << oneRoundGenOp;
+    return oneRoundGenOp;
 }
 
-int CopySetScheduler::DoCopySetSchedule(PoolIdType lid) {
-    // 1. collect the chunkserver list and copyset list of the cluster, then
-    //    collect copyset on every online chunkserver
-    auto copysetList = topo_->GetCopySetInfosInLogicalPool(lid);
-    auto chunkserverList = topo_->GetChunkServersInLogicalPool(lid);
-    std::map<ChunkServerIdType, std::vector<CopySetInfo>> distribute;
-    SchedulerHelper::CopySetDistributionInOnlineChunkServer(
-        copysetList, chunkserverList, &distribute);
-    if (distribute.empty()) {
-        LOG(WARNING) << "no not-retired chunkserver in topology";
-        return UNINTIALIZE_ID;
+int CopySetScheduler::PenddingCopySetSchedule(const std::map<ChunkServerIdType,
+                                    std::vector<CopySetInfo>> &distribute) {
+    int oneRoundGenOp = 0;
+    // for every chunkserver, find one copy set to migrate out
+    for (auto it = distribute.begin(); it != distribute.end(); it++) {
+        ChunkServerIdType source = it->first;
+        int copysetNum = it->second.size();
+        if (copysetNum == 0) {
+            continue;
+        }
+
+        // find one copy set to migrate out from source chunkserver
+        for (auto info : it->second) {
+            // does not meet the basic conditions
+            if (!CopySetSatisfiyBasicMigrationCond(info)) {
+                continue;
+            }
+
+            auto target = SelectBestPlacementChunkServer(info, source);
+            if (target == UNINTIALIZE_ID) {
+                LOG(WARNING) << "copysetScheduler can not select chunkServer "
+                                "to migrate " << info.CopySetInfoStr()
+                            << ", which replica: " << source << " is pendding";
+                continue;
+            }
+
+            Operator op = operatorFactory.CreateChangePeerOperator(
+                info, source, target, OperatorPriority::HighPriority);
+            op.timeLimit = std::chrono::seconds(changeTimeSec_);
+
+            if (AddOperatorAndCreateCopyset(op, info, target)) {
+                oneRoundGenOp++;
+                break;
```

Comment:
why break? One record at a time for a chunkserver? It Should be based on concurrency.

Reply:
done